drivers/aip31068/aip31068.c

```diff
@@ -0,0 +1,516 @@
+}
+
+int aip31068_set_auto_scroll_enabled(aip31068_t *dev, bool enabled)
+{
+    if (enabled) {
+        dev->_curr_entry_mode_set |= (1 << BIT_ENTRY_MODE_AUTOINCREMENT);
+    }
+    else {
+        dev->_curr_entry_mode_set &= ~(1 << BIT_ENTRY_MODE_AUTOINCREMENT);
+    }
+
+    return _command(dev, CMD_ENTRY_MODE_SET | dev->_curr_entry_mode_set);
+}
+
+int aip31068_set_cursor_blinking_enabled(aip31068_t *dev, bool enabled)
+{
+    if (enabled) {
+        dev->_curr_display_control |= (1 << BIT_DISPLAY_CONTROL_CURSOR_BLINKING);
+    }
+    else {
+        dev->_curr_display_control &= ~(1 << BIT_DISPLAY_CONTROL_CURSOR_BLINKING);
+    }
+
+    return _command(dev, CMD_DISPLAY_CONTROL | dev->_curr_display_control);
+}
+
+int aip31068_set_cursor_visible(aip31068_t *dev, bool visible)
+{
+    if (visible) {
+        dev->_curr_display_control |= (1 << BIT_DISPLAY_CONTROL_CURSOR);
+    }
+    else {
+        dev->_curr_display_control &= ~(1 << BIT_DISPLAY_CONTROL_CURSOR);
+    }
+
+    return _command(dev, CMD_DISPLAY_CONTROL | dev->_curr_display_control);
+}
+
+int aip31068_set_cursor_position(aip31068_t *dev, uint8_t row, uint8_t col)
+{
+    col = ((row == 0) ? (col | 0x80) : (col | 0xc0));
+    uint8_t data[] = { 0x80, col };
+
+    return _device_write(dev, data, 2);
+}
+
+int aip31068_set_text_insertion_mode(aip31068_t *dev,
+                                     aip31068_text_insertion_mode_t mode)
+{
+    if (mode == RIGHT_TO_LEFT) {
+        dev->_curr_entry_mode_set &= ~(1 << BIT_ENTRY_MODE_INCREMENT);
+    }
+    else {
+        dev->_curr_entry_mode_set |= (1 << BIT_ENTRY_MODE_INCREMENT);
+    }
+
+    return _command(dev, CMD_ENTRY_MODE_SET | dev->_curr_entry_mode_set);
+}
+
+int aip31068_move_cursor_left(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd &= ~(1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_move_cursor_right(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_scroll_display_left(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_SELECTION);
+    cmd &= ~(1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_scroll_display_right(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_SELECTION);
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_set_custom_symbol(aip31068_t *dev,
+                               aip31068_custom_symbol_t custom_symbol,
+                               uint8_t charmap[])
+{
+    uint8_t location = custom_symbol;
+    int rc = _command(dev, CMD_SET_CGRAM_ADDR | (location << 3));
+
+    if (rc < 0) {
+        return rc;
+    }
+
+    int row_count = dev->params.font_size == FONT_SIZE_5x8 ? 8 : 10;
```

Comment:
magic number?

Reply:
If the font size is 5x8 the functions expects 8 rows, if it is 5x10 then it expects 10 rows for a complete character definition.

Reply:
ok!